### PR TITLE
[Efficient Metadata Operations] Add logic in BlobStore to mark a replica as partially sealed when it crosses threshold.

### PR DESCRIPTION
**Pull request by ankagrawal**
This PR implements the logic in BlobStore to mark a replica as partially sealed when it crosses the threshold. Do note that ReplicaStatusDelegate still doesn't do the partial update in HelixPartiticipant, so this doesn't change the end to end logic yet.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (JingQianCloud) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1122]: Can replicaStatusDelegates have 2 and more entries?
If yes, what if one replicaStatusDelegate.seal is true but the other replicaStatusDelegate.seal is false?

The old code probably has the same code logic. Just want to understand it.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java]: removed.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1120]: So replicaStatusDelegates can have 2 entries during zk migration etc. Out strategy during zk migration was that we will use the replicaStatusDelegates for both the zk endpoints/clusters, and let them sync up for a few days before decommissioning the older zk.

Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1120]: So whether a replica is in sealed/partially_sealed/unsealed state is determined by only three factors
1. Whether the size of used capacity of a replica's log exceeds a certain threshold.
2. Whether there is any other replica of the same partition marked as sealed in helix, so the replica gets a notification from helix.
3. Whether the partition id of the replica is present in partitionOverrideList (store in helix)

In all the cases, if there is any divergence, then it's safe for a replica to temporarily assume the most strict seal state. With compaction and writes, each replica will keep re-evaluating its state and eventually come back to a state that is guided by its log usage.

Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1120]: In this `BlobStore` logic, the only thing that matters for a replica to mark itself as sealed/partially-sealed/unsealed is the log size alone (factor 1 above).

Comment (JingQianCloud) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1120]: Thanks for the explanation.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (JingQianCloud) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:214]: Currently we haven't fully supported the partially seal yet.
After we merge this code, how do we make sure there is no impact to the existing code logic?

Comment (JingQianCloud) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:214]: What's the default value of partialSealThresholdBytesLow and storePartialWriteToReadWriteEnableSizeThresholdPercentageDelta we'll set?

Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:214]: The default value for `storePartialWriteEnableSizeThresholdPercentage` is 50%. And the default value for `storePartialWriteToReadWriteEnableSizeThresholdPercentageDelta` is 5%. They are defined in StoreConfig.java.

Comment (ankagrawal) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:214]: > After we merge this code, how do we make sure there is no impact to the existing code logic?

If you check the `ReplicaStatusDelegate::partialSeal()` implementation, it does the same thing that `ReplicaStatusDelegate::unseal()` does. So effectively, anything that is not sealed will remain not sealed.

Comment (JingQianCloud) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:214]: I think after we set the partialSeal(), even the implementation is to set as unseal, the metrics will start to show as partial sealed.

I'm wondering if we set storePartialWriteEnableSizeThresholdPercentage to storeReadOnlyEnableSizeThresholdPercentage, looks like we won't generate partial seal status at all.

I leave it to you.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (Arun-LinkedIn) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1122]: Should this be `replicaId.isSealed()` since `replicaSealStatus.get()` is local value?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (Arun-LinkedIn) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1120]: nit: would name like `oldReplicaSealStatus` be better?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (Arun-LinkedIn) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1122]: Yeah, I think this should be `replicaId.isSealed()` since `replicaSealStatus` is only updated during two ZK clusters case

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java
Comment (Arun-LinkedIn) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:1182]: nit: would it be okay to rename `sealStatus1` to `sealStatus` and `sealStatus` to `globalSealStatus` or something like that?

**Review by Arun-LinkedIn — APPROVED** [review]

**Pull request merged**
